### ISSUE TITLE
Help Up Next To feature

### ISSUE THREAD

**Issue by JumBay** [trakt/api-help#102]
Hi,

I've managed to make an `Up Next To Watch` feature that works, but I think I'm making too many HTTP requests if the user has a lot of shows in its history. I'd like to know if there is a better way to do so. 
Here is my process:

1. Get user calendars shows for the next 6 days (GET https://api.trakt.tv/calendars/my/shows/2019-08-25/6). If a show has aired today, I clear the cache.
2. Then I call GET https://api.trakt.tv/users/me/watched/shows?extended=full to get all shows history. I check if there are episodes to watch by comparing the number of aired episodes via `result.show.aired_episodes` and the total watched episodes via a loop on `result.seasons`. 
3. Then for each show with episodes to watch, I call GET https://api.trakt.tv/shows/trakId/progress/watched to know information about the next episode to watch.

Of course, this is saved in the cache, but every time an episode is marked as watched, the cache is cleared and the whole process has to be done again.

Thanks for your help.

**Comment by rectifyer** [comment]
Does your app cache show/season/episode info? If so, you could also do the progress calculation locally in the app. You can use the local show/season/episode info and then the output from https://api.trakt.tv/users/me/watched/shows?extended=full to determine the progress.

**Comment by JumBay** [comment]
Yes, I do cache everything, but as I said the cache is cleared every time an episode is marked as watched because of `last_activity` that changes. For instance if someone has 200 shows in his history, I'll have to make 200 HTTP queries

**Comment by rectifyer** [comment]
Check out https://trakt.docs.apiary.io/#reference/shows/updates/get-recently-updated-shows which is a better way to look for show updates. That should cut down significantly on the number of show refreshes needed. Make sure to save the `updated_at` timestamp and you can use that for an additional local comparison before re-caching a show's info.

**Comment by JumBay** [comment]
Ok I’ll have a look. 
Thx

**Comment by JumBay** [comment]
I had a look at this route, it's fine if you check it often, but if the user didn't check it in the last 10 days, for example, it could return too many results.  
It will be nice to have the same route but based on the auth user, or in the `last_activity` route get items ids of what has changed.

**Comment by nixgates** [comment]
@JumBay not sure if you ever got this sorted, or how your caching is set out.

Seren uses the same approach Justin mentioned and polls the shows/updates/ endpoint with a large limit. Once it has the initial page, it inspects the response headers for the page count and then threads the rest of the required requests.

Just did a test of this now, using a large page limit of 500 and the date set so it returned over a weeks worth of data, it returned a total count of 12 pages.

This should heavily reduce the amount of calls you need to make to update the cache in a reactive manner.

**Comment by JumBay** [comment]
@nixgates, as mentioned above, this route could return a lot of data since it's not linked to the auth user. It may reduce the number of calls but it's still not perfect especially on a mobile device. Since the Up Next feature is a have to feature, I think this should be handled by trakt itself for performance reason and to avoid reinventing the wheel.
